### PR TITLE
Fix mkdir autocomplete

### PR DESCRIPTION
**Pull request by drew-gross**
My noobishness at TypeScript is showing through here, I've used a very inelegant way of removing `undefined`s from an array. I would appreciate suggestions for how to improve that.

Thank you for adding me as a Collaborator! Would you like me to merge my own PRs? Merge other people's PRs? Anything else?

### CODE REVIEW

**Comment by ekaragodin** [comment]
Hi, you can use filter method
```javascript
return results.filter(result => result);
```

**Comment by drew-gross** [comment]
@ekaragodin that doesn't work. I change the code to 

`const filteredResults: Suggestion[] = results.filter(result => result);`

and get error:

```
src/plugins/autocompletion_providers/Mkdir.ts(81,15): error TS2322: Type '(Suggestion | undefined)[]' is not assignable to type 'Suggestion[]'.
  Type 'Suggestion | undefined' is not assignable to type 'Suggestion'.
    Type 'undefined' is not assignable to type 'Suggestion'
```

**Comment by ekaragodin** [comment]
```javascript
return flagDescriptions
.map(descriptions => {
// ...
})
.filter(result => result);
```

**Comment by drew-gross** [comment]
That just gives a slightly more confusing error:

```
src/plugins/autocompletion_providers/Mkdir.ts(55,11): error TS2322: Type '() => (Suggestion | undefined)[]' is not assignable to type '(context: AutocompletionContext) => Suggestion[] | Promise<Suggestion[]>'.
  Type '(Suggestion | undefined)[]' is not assignable to type 'Suggestion[] | Promise<Suggestion[]>'.
    Type '(Suggestion | undefined)[]' is not assignable to type 'Promise<Suggestion[]>'.
      Property 'then' is missing in type '(Suggestion | undefined)[]'.
```

**Comment by ekaragodin** [comment]
I use this pattern here and it works https://github.com/shockone/black-screen/blob/master/src/plugins/autocompletion_providers/Vagrant.ts#L7

**Comment by shockone** [comment]
You're welcome. I don't expect you do to anything in particular, it's just if you guys help develop the project then you're collaborators, right?

At work we have this process of peer review, which works pretty well. So, you can merge your PR after somebody has reviewed it. Although, I'd like to review all the PRs at the beginning, if you don't mind, until you have more context.

File: src/plugins/autocompletion_providers/Mkdir.ts [hunk 1]
```diff
@@ -52,16 +52,40 @@ exec("man mkdir", {}, (error: any, stdout: string, stderr: string) => {
...
-        let shortFlagWithoutArgument = description[0].match(/^ *-(\w) *(.*)$/);
-        if (shortFlagWithArgument) {
-            flagMap[shortFlagWithArgument[1]] = combineManPageLines(description.slice(1));
-        } else if (shortFlagWithoutArgument) {
-            flagMap[shortFlagWithoutArgument[1]] = combineManPageLines([shortFlagWithoutArgument[2], ...description.slice(1)]);
-        }
-    });
-    const optionsProvider = mapObject(flagMap, (option, descriptions) => mapSuggestions(shortFlag(option), suggestion => suggestion.withDescription(descriptions)));
-    PluginManager.registerAutocompletionProvider("mkdir", combine([...optionsProvider, directoriesSuggestionsProvider]));
+    const optionsProvider: AutocompletionProvider = () => {
```
Comment (shockone) [inline src/plugins/autocompletion_providers/Mkdir.ts]: You'd need to add the `async` keyword before the parenthesis.

**Comment by drew-gross** [comment]
@shockone sure, I will make no changes to my behaviour for now then :)

@ekaragodin I think that solution isn't working for me because the return type is more complex. I'm not 100% sure though.

**Comment by shockone** [comment]
I think the explicit type annotation is what prevents the compiler to infer the type correctly. In any case, those changes are sufficient to make it compile without errors

```diff
diff --git a/src/plugins/autocompletion_providers/Mkdir.ts b/src/plugins/autocompletion_providers/Mkdir.ts
index 3acf19a..77a1566 100644
--- a/src/plugins/autocompletion_providers/Mkdir.ts
+++ b/src/plugins/autocompletion_providers/Mkdir.ts
@@ -52,8 +52,8 @@ exec("man mkdir", {}, (error: any, stdout: string, stderr: string) => {

     // Extract the paragraphs that describe flags, and parse out the flag data
     let flagDescriptions = manDescriptionParagraphs.filter(lines => lines.length > 0);
-    const optionsProvider: AutocompletionProvider = () => {
-        const results: (Suggestion|undefined)[] = flagDescriptions.map(descriptions => {
+    const optionsProvider: AutocompletionProvider = async() => {
+        return flagDescriptions.map(descriptions => {
             let shortFlagWithArgument = descriptions[0].match(/^ *-(\w) (\w*)$/);
             let shortFlagWithoutArgument = descriptions[0].match(/^ *-(\w) *(.*)$/);
             if (shortFlagWithArgument) {
@@ -77,14 +77,7 @@ exec("man mkdir", {}, (error: any, stdout: string, stderr: string) => {
                     description,
                 });
             }
-        });
-        const filteredResults: Suggestion[] = [];
-        results.forEach(result => {
-            if (result) {
-                filteredResults.push(result);
-            }
-        });
-        return filteredResults;
+        }).filter(result => result);
     };
     const allOptions = [optionsProvider, directoriesSuggestionsProvider];
     PluginManager.registerAutocompletionProvider("mkdir", combine(allOptions));
```

**Pull request merged**